### ISSUE TITLE
[HOLD for payment 2023-11-01] [$1000] Web - LHN still displays "Deleted message" although we deleted comment and its thread completely

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#23450]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to any chat report
2. Type a comment
3. Click "Reply on thread" context menu of this comment
4. Type a reply in that thread
5. Remove the parent comment of that thread
6. Verify that the LHN is displays "Deleted message" in original chat report, and in original chat report, we can still see the comment in step 2, with marked as deleted message
7. Delete the reply comment in step 4
8. Go to the original chat report, the comment in step 2 is disappear

## Expected Result:
 LHN should show the latest visible message.

## Actual Result:
LHN still displays "Deleted message" although we deleted comment and its thread completely, and the comment is already disappear in the chat.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.43-7
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/420787f4-fd3d-442f-aefc-cd0c6d5cb727

https://github.com/Expensify/App/assets/93399543/4c0c0520-72aa-4aca-a313-7b5f483d1fc7

**Expensify/Expensify Issue URL:**
**Issue reported by:** @hoangzinh
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1689983918865759

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01746b90b47c15e650</li>
        <li>Upwork Job ID: 1694864004183359488</li>
        <li>Last Price Increase: 2023-09-15</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Reporter | 26781400</li>
<li>rojiphil | Contributor | 27299208</li>
</ul></li>
    </ul>
</details>

**Comment by s-alves10** [comment]
It looks like this is an intentional behavior, but I think this is an inconsistency.

**Comment by johncschuster** [comment]
@cead22 / @marcaaron it looks like you are both assigned the above-linked issue. Can you investigate whether the linked PR caused a regression?

**Comment by marcaaron** [comment]
See my comment here @johncschuster https://github.com/Expensify/App/issues/19722#issuecomment-1648818264

**Comment by johncschuster** [comment]
Awesome. Thanks, @marcaaron! In that case, should we put this one on hold and wait until #19722 has been resolved to see if this behavior is still an issue, or are you confident #19722 will resolve this issue? If yes, I would close this issue.

**Comment by marcaaron** [comment]
Tbh I'm only involved because I am the deployer this week and something loosely related needed to be reverted to unblock the deploy. I would suggest seeking out and asking the original authors who introduced the regression directly.

**Comment by marcaaron** [comment]
HOLD does sound like a reasonable plan to me though since deleting the comment thread is not that critical of an action that people will take and doesn't have a great effect on the biz.

**Comment by s-alves10** [comment]
https://github.com/Expensify/App/issues/19722#issuecomment-1649225218

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The LHN keeps showing `[Deleted message]` without getting updated with the immediately previous visible message when the Parent Message comment for a threaded chat is deleted.

### What is the root cause of that problem?

The root cause of the problem has to do starting from #19722. Here is a little bit of history of the implementations done before at #19722 for reference. #19722 was implemented using `isLastMessageDeletedParentAction` and specifically this condition `reportAction.childVisibleActionCount > 0` through the [PR implementation for this issue](https://github.com/Expensify/App/pull/22864/files#) to resolve this problem. But, this had caused a blocker issue #23401 which was temporarily unblocked by replacing the condition with `childVisibleActionCount === 0` in [code here](https://github.com/Expensify/App/blob/024d210bc0b5a0454b62a72de60057af67cdd045/src/libs/actions/Report.js#L902). However, as [mentioned in a comment here](https://github.com/Expensify/App/issues/19722#issuecomment-1648818264), the issue is not done yet. This issue occurs when the number of visible actions in the thread chat (Chat T) becomes 0. In this case, `isLastMessageDeletedParentAction` is set because of which the LHN keeps showing `[Deleted message]` without getting updated with the immediately previous visible message. The workaround to overcome this issue is to add a new comment which will reset the variable.

### What changes do you think we should make in order to solve the problem?

Instead of introducing `isLastMessageDeletedParentAction`, we can rely on the combination of existing variables: `isDeletedParentAction` and `childVisibleActionCount`. Since deleting a parent action is a one-way operation, it simplifies the complexity too. We can implement this as follows:

1) Currently, we set `isDeletedParentAction` to true [in this code here](https://github.com/Expensify/App/blob/8cdb506e97e8382f4bc55b0f48784f576fbb2789/src/libs/actions/Report.js#L882) when we delete a comment in parent chat (say Chat P) for which there are visible children in the chat thread (say Chat T). Once this is set, whenever the LHN option row is populated, we can check if this is a deleted parent action and return [Deleted message] as follows. We can implement this just before we check for an empty `lastMessageText` [here in this code](https://github.com/Expensify/App/blob/8cdb506e97e8382f4bc55b0f48784f576fbb2789/src/libs/SidebarUtils.js#L314C11-L314C11) so that the text 'This is the beginning of..' is avoided. 

```
 else {
        const lastAction = visibleReportActionItems[report.reportID];
        if (lodashGet(lastAction, 'actionName', '') === CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT && ReportActionsUtils.isDeletedParentAction(lastAction)) {
            // Called when Parent Report has a most recent visible action that has a thread which is deleted but has more than one visible child comments.  
            result.alternateText = Localize.translate(preferredLocale, 'parentReportAction.deletedMessage');
        }
      }
```

2) As we have already set the `alternateText` to the desired value, we can [replace this line](https://github.com/Expensify/App/blob/8cdb506e97e8382f4bc55b0f48784f576fbb2789/src/components/LHNOptionsList/OptionRowLHN.js#L211) with the following and free ourselves from `isLastMessageDeletedParentAction` by depending directly on `alternateText`. Also, we can remove other references of `isLastMessageDeletedParentAction` in the code.

`{optionItem.alternateText}`

3) Additionally, there is one more bug that needs to be fixed in `deleteReportComment` [here in this code](https://github.com/Expensify/App/blob/8cdb506e97e8382f4bc55b0f48784f576fbb2789/src/libs/actions/Report.js#L965C1-L969C6) . When the Parent Message is deleted in the Thread Chat, the `childVisibleActionCount` will also be decremented which will cause an issue where there is only one visible child action in the Thread Chat. In that case, the LHN will start displaying `[Deleted message]` even though it is not deleted. This can be overcome by checking if the current deleted comment is a parent action as shown below.

```
+    // Update optimistic data for parent report action if the report is a child report and the reportAction is that of the child report.
+    const currentReport = ReportUtils.getReport(reportID);
+    if(currentReport && currentReport.parentReportActionID &&  currentReport.parentReportActionID !== reportAction.reportActionID)
+     {
         const optimisticParentReportData = ReportUtils.getOptimisticDataForParentReportAction(reportID, optimisticReport.lastVisibleActionCreated, CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE);
         if (!_.isEmpty(optimisticParentReportData)) {
             optimisticData.push(optimisticParentReportData);
         }
+     }
```
Regarding BE, this proposal does not depend on `isLastMessageDeletedParentAction` and can be removed. Further, as the FE does not depend on the `lastMessageText` in report for display on deletion of the parent action, BE can behave as it is currently.

Note: Apart from this, there are refreshing issues on LHN which, I believe, are addressed in #23424. 

### What alternative solutions did you explore? (Optional)

**Comment by johncschuster** [comment]
@flodnv @rushatgabhane @dukenv0307, it looks like the PR linked to [this issue](https://github.com/Expensify/App/issues/19722) may have caused a regression. Can you investigate that and leave a comment here so we know how best to handle this issue? Should we close it in favor of #19722?

**Comment by rushatgabhane** [comment]
@johncschuster yes we can hold this issue on #19722  and make it weekly or something

**Comment by johncschuster** [comment]
Sweet! Thanks, @rushatgabhane!

**Comment by chiragsalian** [comment]
I'm still on parental leave and will be back on august 17th. I just wanted to point one thing out in the proposal [here](https://github.com/Expensify/App/issues/23450#issuecomment-1658337106),

> we can rely on the combination of existing variables: isDeletedParentAction and childVisibleActionCount

Feel free to correct me if I'm mistaken (or if things changed recently), but I'm pretty sure you cannot count on `isDeletedParentAction` because its in reportAction in onyx. When the user logs into the app, only onyx report data is up to date and reportAction is not up to date unless the user visits the chat report.

One way to test if your solution works is,
1. Apply your code solution.
2. Assume chat report 111 now shows `[Deleted Message]` in the LHN.
3. Click another chat report say 222 and leave some message.
4. Log out and log in to clear onyx.
5. Once logged in you should be seeing chat report 222 by default. 
6. At this point does chat report 111 show `[Deleted Message]` in its LHN preview? I have a feeling it will not since it won't fetch the updated reportActions just yet.

**Comment by rojiphil** [comment]
> but I'm pretty sure you cannot count on `isDeletedParentAction` because its in reportAction in onyx. When the user logs into the app, only onyx report data is up to date and reportAction is not up to date unless the user visits the chat report.

@chiragsalian  Thanks for your feedback. Please find attached a video with the test case run on the proposed proposal. The test runs well as expected and shows [Deleted message] even after logout and login. Also, the BE is sending the `isDeletedAction` and can be verified in API response. 

I think, before you went on leave, `isDeletedParentAction` was not used effectively as it was always set to true. This was fixed sometime during early July as can [be seen here](https://github.com/Expensify/App/commit/e46cd942e171d51f92f82262c574783b9373a514). Maybe, that's how you arrived at  your understanding.


https://github.com/Expensify/App/assets/3004659/70d445e3-05c1-422d-a400-ad5c13bdfd35

**Comment by johncschuster** [comment]
Bumping to keep Melvin happy while we wait for our friend @chiragsalian to pop back in.

**Comment by chiragsalian** [comment]
im back from parental leave. Also i thought some other internal engineer was helping out here. i didn't realize there was no movement here 😅

Anyway, thanks for the video @rojiphil. That's interesting. I did not think that would work. It looks like there is some flow loading some reportActions. I still doubt that reportActions is a good source of truth for LHN data.

I just tested today to confirm,
1. In production i logged out and back in to clear onyx.
2. I checked onyx data in Application -> Storage -> Indexed DB -> OnyxDB -> keyvaluepairs
3. For `reportActions_` you can see there are just 13 values
![image](https://github.com/Expensify/App/assets/14356145/dae701fb-5207-47c1-8861-500a33e167d5)
4. But for `report_` (which is what LHN normally uses) you can see i have 377.
![image](https://github.com/Expensify/App/assets/14356145/048c75a6-ee86-4cd7-81ba-cfb3836480aa)

It looks like there is some code that returns a few reportActions_ but as long as its count does not match report_ values it cannot be used as a source of truth for LHN UI. Let me know if that makes sense. Feel free to test yourself, you just need to test on an account that has a lot of chats.

**Comment by johncschuster** [comment]
Thanks for jumping in @chiragsalian!

**Comment by rojiphil** [comment]
> I still doubt that reportActions is a good source of truth for LHN data.

@chiragsalian Welcome back. Will look into this and share my comments today

**Comment by rojiphil** [comment]
> It looks like there is some code that returns a few reportActions_ but as long as its count does not match report_ values it cannot be used as a source of truth for LHN UI. Let me know if that makes sense. Feel free to test yourself, you just need to test on an account that has a lot of chats.

@chiragsalian 

I did a test on an account that has lots of chats. My understanding is that it is not a must for every report to have its matching report action. The attached video shows that the report(1117098819918956) that does not have a matching reportAction is pointing to a reportAction(1992965932620012783) within another report(91639911042078). Further, we are already depending on reportActions for LHN as [seen in code here](https://github.com/Expensify/App/blob/329cac3b49fc47da8eac1b7a74a511171fe22345/src/libs/SidebarUtils.js#L296C9-L296C70). So, I think, it is safe to depend on reportActions here. Please let me know your thoughts.

https://github.com/Expensify/App/assets/3004659/133c6e59-7807-4d4b-8fbb-df09fc280036

**Comment by chiragsalian** [comment]
Okay so to explain my perspective, this was my thought process,

Say we have 5 reports - assume reportIDs - 111, 222, 333, 444, 555
I'm saying, the moment we log in in onyx we are fetching all the reports and not all the reportActions i.e., in onyx we can see,
- report_111, report_222, report_333 and report_444, report_555
but we are not fetching all the report actions, i.e., we might have
- reportActions_111, reportActions_222 and nothing else.

Currently on LHN (i.e., SideBarLinks) we only use report data and not reportAction data for its UI. So for a reportID say 555 if we have report_555 and not have reportActions_555 but update SideBarLinks to use reportActions_555 it won't work well since onyx does not have the data for it (unless if we update backend to send all reportActions which i doubt we'll do since it does not sound feasible)

> The attached video shows that the report(1117098819918956) that does not have a matching reportAction is pointing to a reportAction(1992965932620012783) within another report(91639911042078)

But this point is really interesting. It made me test and wonder why this was happening. So it took me a while and i found the backend code. So every report that we fetch, if it has a parent report action then we fetch the parent report action as well. So all reportActions that onyx initially has are all associated with child reports which are used to load the correct title for child reports since the parent report action is the source of truth.

Okay with that in mind your proposal might just be correct since we should always have the parent report action for a child report. I'll test a bit with the deleted action flow as well and review your proposal today.

**Comment by chiragsalian** [comment]
okay `isLastMessageDeletedParentAction `, is definitely not needed since we always have the complete report Actions for a parent report action.

As for your proposal @rojiphil, it mostly looks good to me. There is a lot there. I think breaking each into a separate problem/solution would help with readability but from a high level it makes sense to me.

@johncschuster, I'm marking this as external so that a contributor can help out reviewing the proposal.

**Comment by chiragsalian** [comment]
i removed HOLD since i think thats not needed either

**Comment by johncschuster** [comment]
Thanks @chiragsalian!

**Comment by rojiphil** [comment]
> As for your proposal @rojiphil, it mostly looks good to me. There is a lot there. I think breaking each into a separate problem/solution would help with readability but from a high level it makes sense to me.

@chiragsalian  Thanks for your input. Based on the latest version, I had already broken up into separate problem/solution[ here](https://github.com/Expensify/App/issues/19722#issuecomment-1691484903). Awaiting review inputs from @rushatgabhane .

**Comment by johncschuster** [comment]
Not overdue, Melvin. We're waiting on input from @rushatgabhane

**Comment by johncschuster** [comment]
Bumping in Slack

**Comment by johncschuster** [comment]
In that case, can we close this one out?

**Comment by rojiphil** [comment]
> In that case, can we close this one out?

@johncschuster I have opened a discussion [here in slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1693976699584129) to resolve this.

**Comment by johncschuster** [comment]
Thanks, @rojiphil! I'll review that thread shortly.

**Comment by rojiphil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - LHN still displays "Deleted message" although we deleted the comment and its thread completely

### What is the root cause of that problem?
[Deleted message] gets displayed in LHN as subtitle when `childVisibleActionCount` is 0 [as shown here](https://github.com/Expensify/App/blob/662af46c49bbe9690de71ecfe9ad98a9481f0f24/src/libs/actions/Report.js#L949C1-L955C6) due to which `isLastMessageDeletedParentAction` is set. This causes the LHN to keep showing [Deleted message] because of this [code here](https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/LHNOptionsList/OptionRowLHN.js#L235) This results in the previous message not getting updated.  

### What changes do you think we should make in order to solve the problem?

To solve this problem, two changes can be made:

1) We can avoid setting the `isLastMessageDeletedParentAction` to true by replacing the [code here](https://github.com/Expensify/App/blob/662af46c49bbe9690de71ecfe9ad98a9481f0f24/src/libs/actions/Report.js#L949C1-L955C6) with the following condition:

`if (reportAction.childVisibleActionCount !== 0) {`

2) Further, we can [replace this line here](https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/LHNOptionsList/OptionRowLHN.js#L235) with the following as `isLastMessageDeletedParentAction` is not set anymore.

`{optionItem.alternateText}`

### What alternative solutions did you explore? (Optional)

**Comment by chiragsalian** [comment]
Friendly bump @rushatgabhane, would you be able to review the proposal [here](https://github.com/Expensify/App/issues/23450#issuecomment-1715164690).

**Comment by rojiphil** [comment]
## Proposal

[Updated](https://github.com/Expensify/App/issues/23450#issuecomment-1715164690) 

@allroundexperts 
Added more context to the proposal. Also, added step 3 for some additional code cleanup. 
Kindly review.

**Comment by rojiphil** [comment]
@allroundexperts [That proposal](https://github.com/Expensify/App/issues/23450#issuecomment-1658337106) had references to other issues. That’s why I left it there. Anyway, I have marked it outdated in this context to take this forward.

**Comment by rojiphil** [comment]
@allroundexperts Updated the [proposal](https://github.com/Expensify/App/issues/23450#issuecomment-1715164690) to add more context in alternate solution section. Hope that too helps.

**Comment by rojiphil** [comment]
> @rojiphil Removing [this](https://github.com/Expensify/App/blob/b11bddc054f54bb46d5fb5aaf05e25b8a7df7faf/src/components/LHNOptionsList/OptionRowLHN.js#L235) would cause the `Deleted` text to never show up in LHN. Do we want to do that considering we might have a case where the parent action is deleted but the child count is not zero?

@allroundexperts We have comprehensively covered the topic of LHN subtitle display of `[Deleted message]` in the [proposal here](https://github.com/Expensify/App/issues/19722#issuecomment-1720740066). Maybe, we should wait until that is done. What are your thoughts?

**Comment by situchan** [comment]
Is this still reproducible?

**Comment by situchan** [comment]
@hoangzinh can you please retest latest app?

**Comment by rojiphil** [comment]
@situchan The issue is reproducible in another way. But, the RCA is the same. Please refer to the test video in [Slack here](https://expensify.slack.com/archives/C01GTK53T8Q/p1693977001765749?thread_ts=1693976699.584129&cid=C01GTK53T8Q).

**Comment by johncschuster** [comment]
bumping to keep Melvin happy while the discussion continues

**Comment by johncschuster** [comment]
@allroundexperts where did we land in the discussion above? What do we need to move this forward?

**Comment by rojiphil** [comment]
> @rojiphil Do you have anything to say [here](https://github.com/Expensify/App/issues/23450#issuecomment-1741030716)?

@allroundexperts 
Looking into this today. Will share my findings.

**Comment by rojiphil** [comment]
> @rojiphil Do you have anything to say [here](https://github.com/Expensify/App/issues/23450#issuecomment-1741030716)?

I was wrong in my observation of your analysis [here](https://github.com/Expensify/App/issues/23450#issuecomment-1734654947). Sorry about that. If you observe the attached test video here for the mentioned steps, the deletion of the thread title will show previous visible message which is what we are fixing in [PR here](https://github.com/Expensify/App/pull/28364) for #19722 . 

Our current issue can be reproduced [like this here](https://expensify.slack.com/archives/C01GTK53T8Q/p1693977001765749?thread_ts=1693976699.584129&cid=C01GTK53T8Q). And, I could reproduce this in the latest main.
I will spend some more time on this and update you by Monday.

https://github.com/Expensify/App/assets/3004659/730e9f88-0bc9-4e52-a594-fdcb7ac01260

**Comment by rojiphil** [comment]
> I will spend some more time on this and update you by Monday.

@allroundexperts

I think, the fix to the issue is like this attached video below as compared to the issue [described here](https://expensify.slack.com/archives/C01GTK53T8Q/p1693977001765749?thread_ts=1693976699.584129&cid=C01GTK53T8Q). 
I also think that the proposal as [mentioned here](https://github.com/Expensify/App/issues/23450#issuecomment-1715164690) would fix this.
Hope this brings more clarity on the issue and the required fix. 
Let me know your thoughts on this.

https://github.com/Expensify/App/assets/3004659/283405c7-7b32-455f-97da-91a30d94f00d

**Comment by johncschuster** [comment]
@allroundexperts what do you think about the above?

**Comment by rojiphil** [comment]
@allroundexperts 
Gentle bump on reviewing [this comment](https://github.com/Expensify/App/issues/23450#issuecomment-1763472921)

**Comment by johncschuster** [comment]
bumped in Slack

**Comment by chiragsalian** [comment]
hmmm, can we just remove `isLastMessageDeletedParentAction`? Did you test the steps from this [PR](https://github.com/Expensify/App/pull/22864). Won't the problem from the PR become an issue again if we remove `isLastMessageDeletedParentAction`?

**Comment by rojiphil** [comment]
> hmmm, can we just remove `isLastMessageDeletedParentAction`? Did you test the steps from this [PR](https://github.com/Expensify/App/pull/22864). Won't the problem from the PR become an issue again if we remove `isLastMessageDeletedParentAction`?

@chiragsalian 

We initially used `isLastMessageDeletedParentAction` when the usage of `isDeletedParentAction` and `childVisibleActionCount` was not there in LHN. But, since then, `isDeletedParentAction` and `childVisibleActionCount` have been used as a single source of truth for thread implementations. Also, the LHN updation logic in FE has been enhanced to use these variables. So, I think it is safe to remove `isLastMessageDeletedParentAction` and fix the issue here.

I just tested the steps from the mentioned PR (attached test video for reference) after removing `isLastMessageDeletedParentAction`. It works well. 
Also, can you please check if `isLastMessageDeletedParentAction` can be removed from BE too? Not sure if there are any special cases that you would identify in BE for which we should test. If so, please let me know and I can test it out in FE too.

https://github.com/Expensify/App/assets/3004659/256f3004-b021-4aff-8e06-cef14796539e

**Comment by chiragsalian** [comment]
Neat, thanks for checking, proposal LGTM feel free to create the PR @rojiphil.

**Comment by rojiphil** [comment]
> If no regressions arise, payment will be issued on 2023-11-01. 🎊

@johncschuster 
Gentle bump on payments here

**Comment by johncschuster** [comment]
### Payment Summary:
_The pull request got merged within 3 working days of assignment, so this job is eligible for a 50% #urgency bonus 🎉_ ([Comment](https://github.com/Expensify/App/issues/23450#issuecomment-1775766755))

@rojiphil requires payment [offer](https://www.upwork.com/nx/wm/offer/27299208) (Contributor) - $1500
@allroundexperts does not require payment (Eligible for Manual Requests) - $1500
@hoangzinh requires payment [offer](https://www.upwork.com/nx/wm/offer/27565581) (Reporter) - $250

**Comment by johncschuster** [comment]
@hoangzinh can you accept the offer on Upwork? I'll get payment issued to you once that's done.

**Comment by johncschuster** [comment]
I've issued payment to @hoangzinh. @allroundexperts, can you complete the BZ checklist when you get a moment? Thank you!

**Comment by JmillsExpensify** [comment]
$1,500 payment approved for @allroundexperts based on [this comment](https://github.com/Expensify/App/issues/23450#issuecomment-1800138620).
